identity-service/src/typed-routes/solana/solanaRelayChecks.ts

```diff
@@ -0,0 +1,349 @@
+import {
+  NATIVE_MINT,
+  ASSOCIATED_TOKEN_PROGRAM_ID,
+  TOKEN_PROGRAM_ID,
+  decodeInstruction,
+  isCloseAccountInstruction,
+  isTransferCheckedInstruction
+} from '@solana/spl-token'
+import {
+  PublicKey,
+  TransactionInstruction,
+  Secp256k1Program
+} from '@solana/web3.js'
+import audiusLibsWrapper from '../../audiusLibsInstance'
+import models from '../../models'
+import { InvalidRelayInstructionError } from './InvalidRelayInstructionError'
+import {
+  decodeAssociatedTokenAccountInstruction,
+  isCreateAssociatedTokenAccountInstruction,
+  isCreateAssociatedTokenAccountIdempotentInstruction
+} from './programs/associatedToken'
+import {
+  decodeClaimableTokenInstruction,
+  isTransferClaimableTokenInstruction
+} from './programs/claimable-tokens'
+import { decodeRewardManagerInstruction } from './programs/reward-manager'
+import config from '../../config'
+
+const MEMO_PROGRAM_ID = 'Memo1UhkJRfHyvLMcVucJwxXeuD728EqVDDwQDxFMNo'
+const CLAIMABLE_TOKEN_PROGRAM_ID: string = config.get(
+  'solanaClaimableTokenProgramAddress'
+)
+const REWARDS_MANAGER_PROGRAM_ID: string = config.get(
+  'solanaRewardsManagerProgramId'
+)
+const JUPITER_AGGREGATOR_V6_PROGRAM_ID =
+  'JUP6LkbZbjS1jKKwapdHNy74zcZ3tLUZoi5QNyVTaV4'
+
+const audioMintAddress: string = config.get('solanaMintAddress')
+const usdcMintAddress: string = config.get('solanaUSDCMintAddress')
+
+const REWARD_MANAGER: string = config.get('solanaRewardsManagerProgramPDA')
+
+const claimableTokenAuthorities = [audioMintAddress, usdcMintAddress].reduce(
+  (acc, mint) => ({
+    ...acc,
+    [mint]: PublicKey.findProgramAddressSync(
+      [new PublicKey(mint).toBytes().slice(0, 32)],
+      new PublicKey(CLAIMABLE_TOKEN_PROGRAM_ID)
+    )[0]
+  }),
+  {} as Record<string, PublicKey>
+)
+
+/**
+ * Only allow the createTokenAccount instruction of the Associated Token
+ * Account program, provided it has matching close instructions.
+ * Close instructions are on the Token Programand are not validated here.
+ */
+const assertAllowedAssociatedTokenAccountProgramInstruction = (
+  instructionIndex: number,
+  instruction: TransactionInstruction,
+  instructions: TransactionInstruction[]
+) => {
+  const decodedInstruction =
+    decodeAssociatedTokenAccountInstruction(instruction)
+  if (
```

Comment:
nit: Easier to follow this if we invert and throw early instead of putting the throw at the end of a long conditional block.

Reply:
I'm thinking of leaving it like this because it more easily allows adding more allowed instructions while excluding the rest and keeps a similar control flow for all of these